### ISSUE TITLE
Use the same CUDA stream for all RPCs within the same dist autograd context

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#57960]
Today, each RPC will grab a stream from pool and run the target function. This works for RPC itself. However, the entire training loop contains forward-backward-opt. We have to make sure that all ops in the three steps are synchronized properly. Individual CUDA RPC stream syncs won't address the problem, as there are multiple RPCs launched separately. We tried to fix some of that problem by enforcing orders in the distributed autograd (see #57792), but it's still not sufficient, as optimizer step also needs to synchronize with the next iteration.

## Pitch

Add an argument to distributed autograd context to enforce all RPCs within the context use the same streams. And we will need a named stream to ensure CUDA stream synchronization across iterations. 

```python
named_streams = {
  torch.device(0) : "train",
  RemoteDevice("worker2/cuda:1" : "train",
}
for _ in next_batch():
  with dist_autograd.context(named_streams=named_streams):   # Better names are appreciated
    # forward
    # backward
    # opt.step
```

**Comment by mrshenli** [comment]
cc @pbelevich  @lw @beauby

**Comment by pritamdamania87** [comment]
In a more general case, does this mean even operations like this might not be synchronized properly?

```
def my_add(rref, a):
  return rref.local_value()  + a

def my_mul(rref, b)
  return rref.local_value() * b

rpc.rpc_sync("w1", my_add, rref, torch.rand(10))
rpc.rpc_sync("w1", my_mul, rref, torch.rand(10))
```

If both RPCs use different streams, wouldn't there be a synchronization issue here too?

**Comment by pritamdamania87** [comment]
> Yep, execution order is NOT guaranteed here. But there are no correctness issue in this example, right, as one is not depending on the result of another?

There is a dependency here. The user first calls rpc_sync to add a value to rref on w1 and then calls rpc_sync to multiply a value to the same rref on 1. From the user's perspective these ops should happen in that order (first add and then mul) since both operations are rpc_sync. The full code could be something like this:

```
def my_add(rref, a):
  return rref.local_value()  + a

def my_mul(rref, b)
  return rref.local_value() * b

rref = rpc.remote("w1", torch.zeros, 10)
rpc.rpc_sync("w1", my_add, rref, torch.rand(10))
rpc.rpc_sync("w1", my_mul, rref, torch.rand(10))
```

**Comment by mrshenli** [comment]
Do you mean this instead?


```python
def my_add(rref, a):
  rref.local_value()  += a

def my_mul(rref, b)
  rref.local_value() *= b

rref = rpc.remote("w1", torch.zeros, args=(10))
rpc.rpc_sync("w1", my_add, args=(rref, torch.rand(10)))
rpc.rpc_sync("w1", my_mul, args=(rref, torch.rand(10)))
```

**Comment by pritamdamania87** [comment]
Ah yes my bad, your example above is what I meant (missed adding the += and *=).

**Comment by mrshenli** [comment]
> I guess its also not limited to RPC sync, the behavior seems unexpected even for RPC asyncs if used like this:

Yep, as `rpc_sync` is `rpc_async` + wait. 


> For example you could have N RPCs invoked on the same node, they all first synchronize with the default stream and then proceed in parallel on their own different streams.

This does not have to be the default stream, right? As long as we pick a specific stream for this synchronization, it should work, and we can also avoid messing up with the ops in the callee foreground user thread. 

> When you call wait() on a CUDA RPC, future block the current thread until GPU processing has finished on the remote end. This might actually introduce a big perf overhead and probably is not in line with the async nature of CUDA.

Right, we discussed about this in early design meetings, and ruled this out. @lw might be able to share more details.

> So basically on the callee side we have a global map of Dict[src, events], so whenever we get an RPC from src, we look up this map for events this RPC should synchronize on.

We might need multiple events per src, as users might use stream context in application code. The caller can get the current stream ID, and send that over to the callee. The callee can enforce one event per `(src, src_stream_id)` pair. 

There can be other perf problems though. Suppose I have the following code:

```python
futs = []
for _ in range(10):
  futs.append(rpc.rpc_async("worker1", func1))

futs[0].wait()
rpc.sync("worker1", func2)
```

The sync call on `func2` is only intended to wait for the first async call. However, if the same `(src, stream)` pair share the same event, the sync call might need to unnecessarily wait for other async rpcs as well, depending on how many async RPCs have finished when the `futs[0].wait()` returns.

**Comment by mrshenli** [comment]
> So now when we do an RPC from w0 to w1 and call wait() on w0, at this point w0 does an RPC to w1 indicating to w1 that future RPCs on w1 from w0 should wait on the RPC corresponding to this wait().

I missed this one.  How do we guarantee that the RPC launched by wait() will be executed after all previous RPCs on w1? With the following code, is it possible that the 10th RPC might need to wait for the 1st RPC, because the wait() RPC can be executed earlier?

```python
futs = []
for _ in range(10):
  futs.append(rpc.rpc_async("worker1", func1))

futs[0].wait()
```

**Comment by pritamdamania87** [comment]
Yes, I meant multiple events per `src`. In the code that you shared, you're referring to the fact that two RPCs might share the same stream on the callee side (since we have a fixed pool) and in that case we end up waiting for the other as well right? I think that should be fine since this applies to other CUDA ops in PyTorch overall where we have a fixed pool, but the user might end up using way more streams resulting in reusing the same stream and some of their ops are serialized instead of parallelized as they expected.

Although, I realized this still doesn't resolve our problem since there could be a scenario like this:

```
def my_add(rref, a):
  rref.local_value()  += a

def my_mul(rref, b):
  rref.local_value() *= b

def my_nested_mul(rref, dst, b):
  rpc_sync(dst, my_mul, args=(rref, b))

rref = rpc.remote("w1", torch.zeros, args=(10))
rpc.rpc_async("w1", my_add, args=(rref, torch.rand(10))).wait()
rpc.rpc_async("w2", my_nested_mul, args=(rref, "w1", torch.rand(10))).wait()
```

Now again in this case the user expected add to happen before `mul`. But since `w1` only keeps track of RPCs coming from `w0`, the RPC coming from "w2" to "w1" will not perform any stream synchronization and result in incorrect behavior. One messy fix would be to keep the history of nested RPCs in some metadata collecting all the nodes touched, so when we run something on the callee we have a list of `src` nodes instead of just the last one.

Although, one thing I'm confused about is the following. My understanding is that when we call `wait()` on the caller, we synchronize with the default stream right? In that case, does that mean anything that runs on the default stream would anyways wait for the RPC to finish on the remote end? Or does the invocation of `rpc_sync` on the caller side not involve any GPU operations and as a result, nothing is synchronized and thats why we don't have this guarantee? If so, would a simpler fix just ensure that on the caller side we first always synchronize on the default stream before launching the actual RPC? I'm guessing we probably didn't do this since this slows down the nested RPC case on a remote node?

**Comment by pritamdamania87** [comment]
> I missed this one. How do we guarantee that the RPC launched by wait() will be executed after all previous RPCs on w1? With the following code, is it possible that the 10th RPC might need to wait for the 1st RPC, because the wait() RPC can be executed earlier?

Yes, you are right about this and it is a problem. I don't think there is a guarantee that the RPC launched by `wait()` will execute after all others. It seems like this contract needs to be somehow enforced on the caller side and not the callee, since that is what happens for CPU execution (see also the nested RPC case above where this breaks).
